53 standardize logging

iameneko opened the pull request:
This pull request introduces a series of enhancements to the ChronoLog project's logging system, as part of Task #53 - Standardize Logging. The key changes include:
- **Library Selection:** spdlog, a high-performance, header-only C++ logging library, has been selected for its efficiency and compatibility.
- **Logger Class Implementation:** A new Logger class, leveraging spdlog, has been implemented. This class supports multiple logging levels, file and console output, and is configured via a central configuration file.
- **Standardized Log Format:** The log format across the project has been standardized, focusing on uniformity and readability. The format encompasses elements such as timestamp, logger name, log level, source file, and message.
- **Severity Levels:** Logs are now categorized into TRACE, DEBUG, INFO, WARNING, ERROR, and CRITICAL levels, allowing for more nuanced and effective monitoring.
- **Configuration Management:** Logging settings are now centralized within a configuration file to streamline log management and customization.

These enhancements aim to improve the logging capabilities of the project, ensuring better modularity, efficiency, and monitoring.

Feedback and further suggestions are welcome.

More details included in the following report:
[_53___Logging_Standardization_Report.pdf](https://github.com/grc-iit/ChronoLog/files/13844021/_53___Logging_Standardization_Report.pdf)

iameneko committed:
BUILD: Add the SPDLOG library to spack.yaml file.

iameneko committed:
FEAT: Log basic implementation with files log.h and log.cpp

iameneko committed:
FEAT: ChronoVisor updated and standardized log.

iameneko committed:
FEAT: ChronoKeeper updated and standardized log

iameneko committed:
FEAT: Client updated and standardized log.

iameneko committed:
FEAT: ChronoCommon updated and standardized log.

iameneko committed:
FEAT: ChronoAPI updated and standardized log.

iameneko committed:
FEAT: ChronoStore updated and standardized log.

iameneko committed:
FEAT: Test updated and standardized log.

iameneko committed:
FIX: Deletion of misleading logs

iameneko committed:
FIX: Deletion of pending old log

iameneko committed:
FEAT: Adapt logging to work with parameters defined in the configuration files.

fkengun commented on ChronoAPI/ChronoLog/src/log.cpp:
```diff
@@ -0,0 +1,108 @@
+    {
+        if(logType == "file")
+        {
+            logger = spdlog::basic_logger_mt(loggerName, location);
```
exception needs to be caught as shown in spdlog's example code.

fkengun commented on default_conf.json.in:
```diff
@@ -26,6 +26,14 @@
+    "Logging": {
+      "log": {
+        "type": "console",
+        "file": "../../logs/chronovisor_logfile.txt",
```
too many relative path could be risky. it is very likely that "../../logs" does not exist. I would say just use the current directory unless the logger code checks for the existence and creates it when it does not exist.

ibrodkin replied:
i agree with Kun. configfile shoiuld have full file path.  full file path in the config file can be generated by the a job  that starts it if needed . for now just use current directory

ibrodkin commented on ChronoAPI/ChronoLog/include/log.h:
1. Please only include the files needed for log.h ( i don't think, memory & stdio are needed )
2. To ensure the singleton pattern copy constructor and assignment operator need to be disabled and destructor needs to be public , otherwise  the client app can't destroy the logger object.
3. In multithreaded application there also should be a static mutex protecting the shared pointer because initialize, getLogger, and changeConfiguration can be called from multiple threads
4. initialize and getLogger should return the pointer to the existing Logger object , not throw exception

iameneko replied:
1. True. solved. 
2. Copy constructor and assignment operator are disabled now. For the destructor, I have chosen to keep the destructor private because the class is managed by a std::shared_ptr. This approach automatically handles the logger's lifecycle, including its destruction when the application terminates or when all references are out of scope. This design aligns with the singleton pattern, ensuring controlled and safe management of the logger's instance without requiring public access to the destructor. In any case, do you prefer to make it public?
3. True. Mutexes are included now. 
4. I do not see the point of returning a pointer. For 'initialize' the method does not return anything. It's designed to set up the logger but not to provide the logger instance. In any case, I added return codes for handling errors in initialization process. For getLogger() method, it returns the logger instance if it exists. If the logger is not initialized (logger is nullptr), the method currently ends the program as it shows that it is wrong configured.

ibrodkin commented on ChronoAPI/ChronoLog/src/log.cpp:
```diff
@@ -0,0 +1,108 @@
+    else
+    {
+        throw std::logic_error("Logger is already initialized");
+    }
```
1. I would prefer if the logger used return codes instead of just throwing exceptions. 
2. in initialize() first check if the logger object exists, and just return, no need to throw exceptions for the case that is valid and functional. Plus it's a good practice to make the if(condition) for the shorter condition outcome first, it makes the code much easier to read, there would be fewer indentation levels 
3.  getLogger() and changeConfiguration() should first check if the object doesn't exist and call initialize to create it  if not; then proceed 
4. mutex protection is needed for the Logger pointer

iameneko replied:
1. Return codes are added now. 
2. True. Already modified. 
3. changeConfiguration() is deleted. No needed in the new implementation. For getLogger(), if will imply passing the configurationManager on each call or adding some extra loggic that i do not find appropiete. Which parameters would you use for called initialize? And how do you access those parameters? I prefer the approach of properly setting things up wiuth initialize() and end the execution otherwise. 
4. True. Added.

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
I suggest that we enforce that full configuration file path  is provided on the command line when starting the ChronoKeeper, ConfigurationManager is initialized first, than the logger is initialized with the real filepath from the configuration file.  
I also suggest that we add IP address of the host to the filename, as there would be many keepers.

iameneko replied:
Question. 
I thought about the option of having multiple keepers and I now current implementation does not support it. I added that in the work report. But when many keepers are working the configuration file will be the same? Same exact parameters are going to be use for all keepers? I thought that for each keeper different file would be used (you can change name then) or that configuration file would be changed in the future and we could adapt naming to the new system. 

In any case, we can include that with no problem. But how? Do we want to be able to add a generic name in the config file and then the log to read the host ip and add it add the end? for example:
config file name: chronovisor_log.txt
after formating: chronovisor_log_127.0.0.1.txt
That is what we want?

fkengun replied:
> But when many keepers are working the configuration file will be the same?

Yes. There is only one configuration file that all components use. That is why we have multiple sections in the file for different components. 

> Do we want to be able to add a generic name in the config file and then the log to read the host ip and add it add the end?

No. In most cases, all daemons log to a shared directory on NFS in a distributed setup. If all keepers use the same file name like `chronokeeper_log.txt`, there will be conflicts. If you do not want to deal with the case with multiple IPs like Ares, you may use hostnames in the log file names.

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -96,12 +103,14 @@ int main(int argc, char**argv)
     if(-1 == service_endpoint_from_dotted_string(datastore_service_ip, datastore_service_port, datastore_endpoint))
     {
-        std::cout << "invalid DataStoreAdmin service address" << std::endl;
+        Logger::getLogger()->error(
```
critical error as we can't proceed in this case

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -117,14 +126,19 @@ int main(int argc, char**argv)
     {
-        std::cout << "invalid KeeperRecordingService  address" << std::endl;
+        Logger::getLogger()->error(
+                "[ChronoKeeperInstance] Failed to start KeeperRecordingService. Invalid endpoint provided.");
```
critical error

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -145,22 +159,23 @@ int main(int argc, char**argv)
     if(nullptr == keeperDataAdminService)
     {
-        std::cout << "ERROR: Keeper failed to create DataStoreAdminService; exiting" << std::endl;
+        Logger::getLogger()->error("[ChronoKeeperInstance] Keeper failed to create DataStoreAdminService exiting");
```
critical error

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -117,14 +126,19 @@ int main(int argc, char**argv)
-    std::cout << keeperIdCard << std::endl;
+
+    std::stringstream ss;
+    ss << keeperIdCard;
```
info level

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -173,29 +188,30 @@ int main(int argc, char**argv)
-
+        std::stringstream s1;
+        s1 << recordingEngine->self();
+        Logger::getLogger()->debug(
```
info

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -173,29 +188,30 @@ int main(int argc, char**argv)
     if(nullptr == keeperRecordingService)
     {
-        std::cout << "ERROR: Keeper failed to create KeeperRecordingService; exiting" << std::endl;
+        Logger::getLogger()->error("[ChronoKeeperInstance] Keeper failed to create KeeperRecordingService exiting");
```
critical

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -206,15 +222,17 @@ int main(int argc, char**argv)
     if(nullptr == keeperRegistryClient)
     {
-        std::cout << "ERROR: Keeper failed to create KeeperRegistryClient; exiting" << std::endl;
+        Logger::getLogger()->error("[ChronoKeeperInstance] Keeper failed to create KeeperRegistryClient; exiting");
```
critical

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -226,22 +244,25 @@ int main(int argc, char**argv)
     {
-        std::cout << "ERROR: Keeper failed to register with the ChronoVisor; exiting" << std::endl;
+        Logger::getLogger()->error(
+                "[ChronoKeeperInstance] Failed to register with ChronoVisor after multiple attempts. Exiting.");
```
critical

ibrodkin commented on ChronoKeeper/ChronoKeeperInstance.cpp:
```diff
@@ -252,28 +273,26 @@ int main(int argc, char**argv)
     delete recordingEngine;
     delete dataAdminEngine;
-
+    Logger::getLogger()->info("[ChronoKeeperInstance] Shutdown completed. Exiting.");
```
Logger object should destroyed , destructor should be public

iameneko replied:
I have chosen to keep the destructor private because the class is managed by a std::shared_ptr. This approach automatically handles the logger's lifecycle, including its destruction when the application terminates or when all references are out of scope. This design aligns with the singleton pattern, ensuring controlled and safe management of the logger's instance without requiring public access to the destructor. In any case, do you prefer to make it public?

ibrodkin commented on ChronoKeeper/DataStoreAdminService.h:
```diff
@@ -27,7 +26,7 @@ class DataStoreAdminService: public tl::provider <DataStoreAdminService>
     ~DataStoreAdminService()
     {
-        std::cout << "DataStoreAdminService::~DataStoreAdminService" << std::endl;
+        Logger::getLogger()->info("[DataStoreAdminService] Destructor called. Cleaning up...");
```
debug not info

ibrodkin commented on ChronoKeeper/IngestionQueue.h:
```diff
@@ -33,61 +34,72 @@ class IngestionQueue
         if(orphanEventQueue.empty())
-        { return; }
-
+        {
```
debug not info

ibrodkin commented on ChronoKeeper/KeeperDataStore.cpp:
```diff
@@ -27,14 +28,15 @@ int chronolog::KeeperDataStore::startStoryRecording(std::string const &chronicle
 {
-    std::cout << "KeeperDataStore: received startStoryRecording {" << chronicle << ", " << story << ", " << story_id
-              << "}" << std::endl;
+    Logger::getLogger()->debug("[KeeperDataStore] Start recording story: Chronicle={}, Story={}, StoryID={}", chronicle
```
info instead of debug

ibrodkin commented on ChronoKeeper/KeeperDataStore.cpp:
```diff
@@ -27,14 +28,15 @@ int chronolog::KeeperDataStore::startStoryRecording(std::string const &chronicle
     auto pipeline_iter = theMapOfStoryPipelines.find(story_id);
     if(pipeline_iter != theMapOfStoryPipelines.end())
     {
+        Logger::getLogger()->debug("[KeeperDataStore] Story already being recorded. StoryID: {}", story_id);
```
info instead of debug

ibrodkin commented on ChronoKeeper/KeeperDataStore.cpp:
```diff
@@ -186,19 +189,24 @@ void chronolog::KeeperDataStore::dataCollectionTask()
     if(is_running() || is_shutting_down())
-    { return; }
+    {
+        Logger::getLogger()->warn(
```
info instead of warning, it's harmless

ibrodkin commented on ChronoKeeper/KeeperDataStore.cpp:
```diff
@@ -211,20 +219,25 @@ void chronolog::KeeperDataStore::startDataCollection(int stream_count)
     if(is_shutting_down())
-    { return; }
-
+    {
```
info instead of warning

ibrodkin commented on ChronoKeeper/StoryChunkExtractionQueue.h:
```diff
@@ -19,15 +20,20 @@ class StoryChunkExtractionQueue
     ~StoryChunkExtractionQueue()
-    { shutDown(); }
+    {
+        Logger::getLogger()->info("[StoryChunkExtractionQueue] Destructor called. Initiating queue shutdown.");
```
debug instead of info

ibrodkin commented on ChronoKeeper/StoryChunkExtractor.cpp:
```diff
@@ -39,29 +43,34 @@ void chronolog::StoryChunkExtractorBase::shutdownExtractionThreads()
     if(extractorState == SHUTTING_DOWN)
-    { return; }
+    {
+        Logger::getLogger()->warn(
```
info instead of warn

ibrodkin commented on ChronoKeeper/StoryPipeline.cpp:
```diff
@@ -100,9 +106,8 @@ void chronolog::StoryPipeline::finalize()
-            std::cout << "StoryPipeline::finalize: storyId { " << storyId << "} extracted chunk {"
-                      << extractedChunk->getStartTime() << "} is empty { "
-                      << (extractedChunk->empty() ? "true" : "false") << std::endl;
+            Logger::getLogger()->debug("[StoryPipeline] Finalized chunk for StoryID={}. Is empty: {}", storyId
```
this is a trace message

ibrodkin commented on ChronoKeeper/StoryPipeline.cpp:
```diff
@@ -128,8 +133,8 @@ std::map <uint64_t, chronolog::StoryChunk*>::iterator chronolog::StoryPipeline::
     std::time_t time_t_chunk_end = std::chrono::high_resolution_clock::to_time_t(chunk_end_point);
-    std::cout <<"StoryPipeline: storyId { " << storyId<<"} prepend chunk {" << timelineStart<<" "<< std::ctime(&time_t_chunk_start)
-    <<"} {" << timelineStart-chunkGranularity <<" " <<std::ctime(&time_t_chunk_end) <<"}"<<std::endl;
+    Logger::getLogger()->debug("[StoryPipeline] Prepending new chunk for StoryID={} starting at: {}",
```
trace

ibrodkin commented on ChronoKeeper/StoryPipeline.cpp:
```diff
@@ -155,8 +160,8 @@ std::map <uint64_t, chronolog::StoryChunk*>::iterator chronolog::StoryPipeline::
     std::time_t time_t_chunk_end = std::chrono::high_resolution_clock::to_time_t(chunk_end_point);
-    std::cout <<"StoryPipeline: storyId { "<< storyId<<"} append chunk {" << timelineEnd <<" "<< std::ctime(&time_t_chunk_start)
-    <<"} {" << timelineEnd+chunkGranularity <<" " <<std::ctime(&time_t_chunk_end) <<"}"<<std::endl;
+    Logger::getLogger()->debug("[StoryPipeline] Appending new chunk for StoryID={} starting at: {}",
```
trace

ibrodkin commented on ChronoKeeper/StoryPipeline.cpp:
```diff
@@ -194,16 +200,17 @@ void chronolog::StoryPipeline::extractDecayedStoryChunks(uint64_t current_time)
-              << std::ctime(&time_t_current_time) << " storyId { " << storyId << "} size {" << storyTimelineMap.size()
-              << "} head_chunk decay_time {" << std::ctime(&time_t_decay) << "}" << std::endl;
+    Logger::getLogger()->debug(
+            "[StoryPipeline] StoryID: {} - Current time: {} - Timeline size: {} - Head chunk decay time: {}", storyId
```
trace instead of debug

ibrodkin commented on ChronoKeeper/StoryPipeline.cpp:
```diff
@@ -218,9 +225,9 @@ void chronolog::StoryPipeline::extractDecayedStoryChunks(uint64_t current_time)
-            std::cout << "StoryPipeline:extractDecayedStoryChunks: storyId { " << storyId << "} extracted chunk {"
-                      << extractedChunk->getStartTime() << "} is empty { "
-                      << (extractedChunk->empty() ? "true" : "false") << std::endl;
+            Logger::getLogger()->debug("[StoryPipeline] StoryID: {} - Extracted chunk with start time {} is empty: {}"
```
trace instead of debug

ibrodkin commented on ChronoKeeper/StoryPipeline.cpp:
```diff
@@ -233,8 +240,8 @@ void chronolog::StoryPipeline::extractDecayedStoryChunks(uint64_t current_time)
 #ifdef TRACE_CHUNK_EXTRACTION
-    std::cout << "StoryPipeline::extractDecayedStoryChunks: storyId { " << storyId << "} extractionQueue.size {"
-              << theExtractionQueue.size() << "}" << std::endl;
+    Logger::getLogger()->debug("[StoryPipeline] Extracting decayed chunks for StoryID={}. Queue size: {}", storyId
```
trace instead of debug

ibrodkin commented on ChronoKeeper/chrono_common/KeeperIdCard.h:
nothing but formatting changes .. i thought we were done with those

ibrodkin commented on ChronoVisor/include/ClientPortalService.h:
```diff
@@ -137,9 +133,11 @@ class ClientPortalService: public thallium::provider <ClientPortalService>
+        //setup finalization callback in case this ser vice provider is still alive when the engine is finalized
+        std::stringstream ss;
+        ss << get_engine().self();
+        Logger::getLogger()->debug("[ClientPortalService] Started at address {} with provider id {}", ss.str()
```
info instead of debug

ibrodkin commented on ChronoVisor/include/DataStoreAdminClient.h:
```diff
@@ -43,22 +43,22 @@ class DataStoreAdminClient
         try
         {
-            std::cout << "DataStoreAdminClient: shutdown_collection:" << std::endl;
+            Logger::getLogger()->debug("[DataStoreAdminClient] Shutdown");
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -32,22 +31,22 @@ int KeeperRegistry::InitializeRegistryService(ChronoLog::ConfigurationManager co
         if(MARGO_INSTANCE_NULL == margo_id)
         {
-            std::cout << "KeeperRegistryService: Failed to initialize margo_instance" << std::endl;
+            Logger::getLogger()->error("[KeeperRegistry] Failed to initialize margo_instance");
```
critical instead of error

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -32,22 +31,22 @@ int KeeperRegistry::InitializeRegistryService(ChronoLog::ConfigurationManager co
         }
-        std::cout << "KeeperRegistryService:margo_instance initialized with NA_STRING" << "{"
-                  << KEEPER_REGISTRY_SERVICE_NA_STRING << "}" << std::endl;
+        Logger::getLogger()->debug("[KeeperRegistry] margo_instance initialized with NA_STRING {}"
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -32,22 +31,22 @@ int KeeperRegistry::InitializeRegistryService(ChronoLog::ConfigurationManager co
-
+        std::stringstream ss;
+        ss << registryEngine->self();
+        Logger::getLogger()->debug("[KeeperRegistry] Starting at address {} with provider id: {}", ss.str()
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -66,14 +65,18 @@ int KeeperRegistry::InitializeRegistryService(ChronoLog::ConfigurationManager co
 
     if(is_shutting_down())
-    { return chronolog::CL_SUCCESS; }
+    {
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -66,14 +65,18 @@ int KeeperRegistry::InitializeRegistryService(ChronoLog::ConfigurationManager co
+    }
 
     registryState = SHUTTING_DOWN;
+    Logger::getLogger()->debug("[KeeperRegistry] Shutting down...");
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -84,19 +87,17 @@ int KeeperRegistry::ShutdownRegistryService()
-                          << std::endl;
+                std::stringstream ss;
+                ss << process.second.idCard;
+                Logger::getLogger()->debug("[KeeperRegistry] Sending shutdown to keeper {}", ss.str());
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -157,25 +160,27 @@ int KeeperRegistry::registerKeeperProcess(KeeperRegistrationMsg const &keeper_re
-              << service_na_string << ": provider_id=" << admin_service_id.provider_id << "}" << std::endl;
+    std::stringstream s1;
+    s1 << keeper_id_card;
+    Logger::getLogger()->debug(
```
error instead of debug

iameneko replied:
why?

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -157,25 +160,27 @@ int KeeperRegistry::registerKeeperProcess(KeeperRegistrationMsg const &keeper_re
     { registryState = RUNNING; }
 
-    std::cout << "KeeperRegistry : RUNNING with {" << keeperProcessRegistry.size() << "} KeeperProcesses" << std::endl;
+    Logger::getLogger()->debug("[KeeperRegistry] RUNNING with {} KeeperProcesses", keeperProcessRegistry.size());
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -206,8 +211,8 @@ int KeeperRegistry::unregisterKeeperProcess(KeeperIdCard const &keeper_id_card)
         registryState = INITIALIZED;
-        std::cout << "KeeperRegistry: state {INITIALIZED}" << " with {" << keeperProcessRegistry.size()
-                  << "} KeeperProcesses" << std::endl;
+        Logger::getLogger()->debug("[KeeperRegistry] INITIALIZED with {} KeeperProcesses"
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -262,7 +267,7 @@ int KeeperRegistry::notifyKeepersOfStoryRecordingStart(std::vector <KeeperIdCard
     if(!is_running())
     {
-        std::cout << "Registry has no Keeper processes to start story recording" << std::endl;
+        Logger::getLogger()->debug("[KeeperRegistry] Registry has no Keeper processes to start story recording");
```
warning instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -299,26 +312,32 @@ int KeeperRegistry::notifyKeepersOfStoryRecordingStart(std::vector <KeeperIdCard
-                      << "} start_time {" << story_start_time << "}" << std::endl;
+            std::stringstream s4;
+            s4 << keeper_id_card;
+            Logger::getLogger()->debug(
```
info instead of debug

ibrodkin commented on ChronoVisor/src/KeeperRegistry.cpp:
```diff
@@ -346,41 +365,49 @@ int KeeperRegistry::notifyKeepersOfStoryRecordingStop(std::vector <KeeperIdCard>
-                      << "}" << std::endl;
+            std::stringstream s3;
+            s3 << keeper_id_card;
+            Logger::getLogger()->debug("[KeeperRegistry] Registry notified keeper {} to stop recording story {}"
```
info instead of debug

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -48,40 +45,34 @@ int chronolog::VisorClientPortal::StartServices(ChronoLog::ConfigurationManager
 
-        std::cout << "VisorClientPortal : :margo_instance initialized with NA_STRING" << "{"
-                  << CLIENT_PORTAL_SERVICE_NA_STRING << "}" << std::endl;
+        Logger::getLogger()->debug("[VisorClientPortal] margo_instance initialized with NA_STRING {}"
```
info instead of debug

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -113,28 +103,29 @@ chronolog::VisorClientPortal::~VisorClientPortal()
                                                 , chl::ClientId &client_id, uint64_t &clock_offset)
 {
-    LOGD("%s called with args: account=%u host_id=%u pid=%u", __FUNCTION__, client_euid, client_host_id, client_pid);
+    Logger::getLogger()->info("New Client Connected.");
```
please include client parameters passed as in the original message

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -113,28 +103,29 @@ chronolog::VisorClientPortal::~VisorClientPortal()
-    LOGD("%s is called with args: client_id=%lu", __FUNCTION__, client_id);
-
+    Logger::getLogger()->info("Client Disconnected.");
+    Logger::getLogger()->debug("[VisorClientPortal] Disconnected client arguments: ClientID={}", client_id);
```
info instead of debug

ibrodkin replied:
one log message of info level with client_id included

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -113,28 +103,29 @@ chronolog::VisorClientPortal::~VisorClientPortal()
     client_id = client_token;    //INNA: change this API...
-    LOGD("%s  with args: account=%u host_id=%u pid=%u -> client_token=%lu", __FUNCTION__, client_euid, client_host_id
-         , client_pid, client_token);
+    Logger::getLogger()->debug("[VisorClientPortal] Client arguments: account={} host_id={} pid={} -> client_token={}"
```
info instead of debug

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -145,40 +136,45 @@ int chronolog::VisorClientPortal::CreateChronicle(chl::ClientId const &client_id
     int return_code = chronicleMetaDirectory.create_chronicle(chronicle_name);
-
+    if(return_code == CL_SUCCESS)
+    {
```
info instead of debug

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -145,40 +136,45 @@ int chronolog::VisorClientPortal::CreateChronicle(chl::ClientId const &client_id
-    LOGD("%s is called in PID=%d, with args: chronicle_name=%s, story_name=%s", __FUNCTION__, getpid()
-         , chronicle_name.c_str(), story_name.c_str());
-
+    Logger::getLogger()->debug("[VisorClientPortal] Story destroyed: PID={}, ChronicleName={}, StoryName={}", getpid()
```
info instead of debug

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -224,7 +217,15 @@ chronolog::VisorClientPortal::AcquireStory(chl::ClientId const &client_id, std::
+    }
+    else
+    {
+        Logger::getLogger()->debug(
```
info instead of debug

ibrodkin commented on ChronoVisor/src/VisorClientPortal.cpp:
```diff
@@ -240,18 +241,16 @@ chronolog::VisorClientPortal::AcquireStory(chl::ClientId const &client_id, std::
 {
-    LOGD("%s is called in PID=%d, with args: chronicle_name=%s, story_name=%s", __FUNCTION__, getpid()
-         , chronicle_name.c_str(), story_name.c_str());
+    Logger::getLogger()->debug("[VisorClientPortal] Story released: PID={}, ChronicleName={}, StoryName={}", getpid()
```
info instead of debug

ibrodkin commented on ChronoVisor/src/chronovisor_instance.cpp:
same comment as i made about ChronoKeeperInstance.ccp. We need to enforce that configuration file path is provided on the command line , configuration manager is the first to be initialized then the logger.

ibrodkin commented on ChronoVisor/src/chronovisor_instance.cpp:
```diff
@@ -36,39 +33,41 @@ int main(int argc, char**argv)
     theChronoVisorPortal.ShutdownServices();
     keeperRegistry.ShutdownRegistryService();
-
+    Logger::getLogger()->info("[chronovisor_instance] ChronoVisor shutdown.");
```
logger object needs to be destroyed, as it's created on the heap

iameneko replied:
I have chosen to keep the destructor private because the class is managed by a std::shared_ptr. This approach automatically handles the logger's lifecycle, including its destruction when the application terminates or when all references are out of scope. This design aligns with the singleton pattern, ensuring controlled and safe management of the logger's instance without requiring public access to the destructor. In any case, do you prefer to make it public?

ibrodkin commented on Client/src/KeeperRecordingClient.h:
```diff
@@ -56,6 +60,7 @@ class KeeperRecordingClient
     ~KeeperRecordingClient()
     {
         record_event.deregister();
+        Logger::getLogger()->info("[KeeperRecordingClient] Destructor called. Deregistered record_event.");
```
debug instead of info

ibrodkin commented on Client/src/StorytellerClient.cpp:
```diff
@@ -203,7 +221,9 @@ chronolog::StorytellerClient::initializeStoryWritingHandle(ChronicleName const &
     if(story_record_iter != acquiredStoryHandles.end())
     {
-        return (*story_record_iter).second;
+        Logger::getLogger()->info("[StorytellerClient] StoryHandle already exists for Chronicle: '{}' and Story: '{}'."
```
debug instead of info

ibrodkin commented on Client/src/StorytellerClient.cpp:
```diff
@@ -140,19 +144,20 @@ int chronolog::StorytellerClient::addKeeperRecordingClient(chronolog::KeeperIdCa
-    std::cout << "StorytellerClient: RUNNING with {" << recordingClientMap.size() << "} KeeperRecordingClients"
-              << std::endl;
-
+    Logger::getLogger()->debug("[StorytellerClient] RUNNING with {} KeeperRecordingClients", recordingClientMap.size());
```
info instead of debug

ibrodkin commented on Client/src/StorytellerClient.cpp:
```diff
@@ -257,8 +288,17 @@ void chronolog::StorytellerClient::removeAcquiredStoryHandle(ChronicleName const
     auto story_record_iter = acquiredStoryHandles.find(std::pair <std::string, std::string>(chronicle, story));
     if(story_record_iter != acquiredStoryHandles.end())
     {
-        delete (*story_record_iter).second;
```
please revert this line

ibrodkin commented on chrono_common/AcquireStoryResponseMsg.h:
formatting changes again...

ibrodkin commented on chrono_common/ConnectResponseMsg.h:
why would you change the ostream operator for all the messages ?   not a big deal I guess , just curious

ibrodkin commented on chrono_common/KeeperIdCard.h:
formatting changes ...

ibrodkin commented on chrono_common/KeeperRegistrationMsg.h:
formatting changes

ibrodkin commented on chrono_common/chronolog_types.h:
why does chronolog_types.h need log.h.  ?

iameneko committed:
FIX: Pending client info with logging connection and desconnection

iameneko committed:
FIX: Reverting unnecessary changes.

iameneko committed:
FIX: Correct log levels

iameneko committed:
FIX: Unnecessary include deleted

iameneko committed:
FIX: Double disconnect request

iameneko committed:
FIX: Revert Unnecessary formatting changes

iameneko committed:
CHORE: Include the log library in required files.

iameneko committed:
FEAT: Change components initialization to require a configuration file in the initialization. Change the log for the configuration file and use a single and personalized logger for the project.

iameneko committed:
FEAT: Add option to disable logging. Configure logger in off mode.

iameneko committed:
FIX: Delete relative paths from config files.

iameneko committed:
FEAT: Redesign of the logger class.
1. Add exception handling in initialize()
2. Add return values in initialize()
3. Change conditions to minimize indention and improve readability.
4. Block trace and debug logging levels on release builds.
5. Add mutexes
6. Enforce the configuration of a logger.
7. Add file rotation functionality.
8. Copy constructor and copy assignment operator have been deleted for singleton pattern.
9. Clean up code.

iameneko committed:
FEAT: Adapt test cases to return values of the initialization method of the log class.

iameneko committed:
FEAT: Addition of MACROS for avoiding trace and debug log level compilation on Release builds

iameneko committed:
FEAT: Load filesize and filenum parameters for the filerotation of the logger from the configuration file.

iameneko committed:
FEAT: Include hostname in the logfile name.

iameneko committed:
DOC: Update Doxygen documentation to match input paramters of the initialize method

iameneko committed:
FIX: Destructor of the logger class is public now.

iameneko committed:
FIX: Revert log filenames changes.

iameneko committed:
FEAT: Addition of consistency to log class.

iameneko committed:
FEAT: Log macros improvement for better understanding

fkengun commented on Client/storyteller_test/storyteller_test.cpp:
```diff
@@ -39,8 +39,10 @@ int main(int argc, char**argv)
         chl::LogEvent event(1, log_time, 7, i, "line_" + std::to_string(i));
-        std::cout << "generated_event {" << event.storyId << ":" << event.time() << ":" << event.clientId << ":"
-                  << event.index() << "}" << std::endl;
+        LOG_DEBUG("generated_event {}:{}:{}:{}", event.storyId, event.time(), event.clientId, event.index());
```
In my case, log.h needs to be included to build this file.

iameneko replied:
Done

iameneko committed:
FIX: Pointer was not initialized when log type was not properly defined.

iameneko committed:
FIX: log.h include was pending on storyteller_test.cpp file.

ibrodkin reviewed (approved):
I'm approving these changes as the task of standardizing of logging in Chronolog code is accomplished.

For the record I think that we'd still tweak the log.h & log.cpp going forward: things like adding time based rotation sink type , removing the lock from Logger::getInstance()to allow this method to be inlined and improve performance, or making console sink the default option, etc...  But these other changes would be localized and as we already have 72 files touched in this Issue implementation, I'd rather have the current implementation merged in and the tweaking postponed.
